crates/matrix-sdk-crypto-nodejs/src/ts/OlmMachine.ts

```diff
@@ -0,0 +1,318 @@
+
+    public static withSledBackend(userId: string, deviceId: string, engine: OlmEngine, sledPath: string) {
+        return (new OlmMachine(userId, deviceId, engine)).makeSled(sledPath);
+    }
+
+    public async runEngineUntilComplete(): Promise<void> {
+        const requests = toArray(this.machine.outgoingRequests ?? []).map((r: string) => JSON.parse(r));
+        for (const request of requests) {
+            await this.runEngineRequest(request);
+        }
+        if ((this.machine.outgoingRequests ?? []).length > 0) {
+            // recurse, hopefully not too much
+            await this.runEngineUntilComplete();
+        }
+    }
+
+    private async runEngineRequest(request: any) {
+        switch(request['request_kind']) {
+            case 'KeysUpload': {
+                const resp = await this.engine.uploadOneTimeKeys(request['body']);
+                this.machine.markRequestAsSent(request['request_id'], RequestKind.KeysUpload, JSON.stringify(resp));
+                break;
+            }
+            case 'KeysQuery': {
+                const userIds = Array.isArray(request['users']) ? request['users'] : [request['users']];
+                const resp = await this.engine.queryOneTimeKeys(userIds);
+                this.machine.markRequestAsSent(request['request_id'], RequestKind.KeysQuery, JSON.stringify(resp));
+                break;
+            }
+            case 'KeysClaim': {
+                const resp = await this.engine.claimOneTimeKeys(request['one_time_keys']);
+                this.machine.markRequestAsSent(request['request_id'], RequestKind.KeysClaim, JSON.stringify(resp));
+                break;
+            }
+            case 'ToDevice': {
+                await this.engine.sendToDevices(request['event_type'], request['body']);
+                break;
+            }
+            default:
+                // TODO: Handle properly
+                console.error("Unhandled request:", request);
+                break;
+        }
+    }
+
+    // --- proxy interface below here ---
+
+    public get userId(): string {
+        return this.machine.userId;
+    }
+
+    public get deviceId(): string {
+        return this.machine.deviceId;
+    }
+
+    public get deviceDisplayName(): Optional<string> {
+        return this.machine.deviceDisplayName;
+    }
+
+    public get identityKeys(): Record<string, string> {
+        return this.machine.identityKeys;
+    }
+
+    public getDevice(userId: string, deviceId: string): Optional<Device> {
+        return this.machine.getDevice(userId, deviceId);
+    }
+
+    public getUserDevices(userId: string): Device[] {
+        return this.machine.getUserDevices(userId);
+    }
+
+    public async pushSync(events: ToDeviceEvent[], deviceLists: DeviceLists, remainingKeyCounts: Record<string, number>, unusedFallbackKeyTypes?: string[]) {
+        const keyCounts = Object.entries(remainingKeyCounts).map(e => [e[0], e[1].toString()]).reduce((c, p) => ({...c, [p[0]]: p[1]}), {});
```

Comment:
napi-rs supports it, but Node doesn't. It couldn't figure out what the numeric type on the JS side was, so ended up wrapping it as a string. 

Will document this in code.

Reply:
Hmm? But napi-rs is made to work with node specifically, seems like `u32` or `i64` would work just fine on the Rust side: https://napi.rs/docs/concepts/values#numbers. The diff I posted earlier uses i64, but maybe using u32 would be better since that can be converted to the `UInt` you need in the end without any error handling.

Reply:
Rust -> JS is fine, but JS -> Rust can't convert a JS `Number` to Rust types at runtime, annoyingly.

Reply:
The linked example seems to expose a method that accepts a numeric type from the js side and returns a number type to the js side:

```rust
#[napi]
fn sum(a: u32, b: i32) -> i64 {
    (b + a as i32).into()
}
```

Which numeric type did you use that failed conversion at runtime?

Reply:
iirc it was an `i32`

from what I remember the error was coming from node rather than napi-rs, so it's possible that napi-rs doesn't really work on certain node versions. Will play with it some more.